716.py

```diff
@@ -0,0 +1,60 @@
+class MaxStack:
+
+    def __init__(self):
+        """
+        initialize your data structure here.
+        """
+        self.stack = []
+        self.max_stack = []
+
+
+    def push(self, x: int) -> None:
+        self.stack.append(x)
+
+        if not self.max_stack or x >= self.max_stack[-1]:
+            self.max_stack.append(x)
+
+
+    def pop(self) -> int:
+        x = self.stack.pop()
+
+        if x == self.max_stack[-1]:
+            self.max_stack.pop()
+
+        return x
+
+
+    def top(self) -> int:
+        return self.stack[-1]
+
+
+    def peekMax(self) -> int:
+        return self.max_stack[-1]
+
+
+    def popMax(self) -> int:
```

Comment:
In this particular design, `popMax()` has a time complexity of `O(N)`. We can do better than this by _not_ using stacks or lists as the underlying data structure. This problem can be as difficult as the [LRU Cache](https://leetcode.com/problems/lru-cache/) problem.